feat(kradio): component reskinning [KHCP-8996] (#1781)
Kong/kongponents: docs/components/radio.md

# Radio

> KRadio is a user-friendly wrapper around radio input element.

Reviewer: ```suggestion
KRadio is wrapper for native input type `radio` elements.
```

Reviewer: Something like this?

### selectedValue

The value of the KRadio option that will be emitted by the `change` and `update:modelValue` events.

### label

Will place label text to the right of the radio. Can also be [slotted](#slots).

<KRadio v-model="labelPropRadio" label="Label example" :selected-value="true" />

```html
<KRadio v-model="checked" label="Label example" :selected-value="true" />
```

### description

Will place description text under the radio label. Can also be [slotted](#slots).

<KRadio v-model="descriptionPropRadio" label="Label example" description="Some description text" :selected-value="true" />

```html
<KRadio
  v-model="radio"
  label="Label example"
  description="Some description text"
  :selected-value="true"
/>
```

### error

Use this prop to apply error styling to the component.

<KRadio v-model="errorPropRadio" label="Input error" error description="Some description text" :selected-value="true" />

```html
<KRadio
  v-model="radio"
  label="Input error"
  error
  description="Some description text"
  :selected-value="true"
/>
```

> ### card

Author: Done

      card
      label="Public"
      selected-value="public"
    >
      <WorldIcon />
    </KRadio>
    <KRadio
      v-model="cardRadio"
      description="Choose this option if you want your APIs to only be accessible from within your private network."
      card
      label="Private"
      selected-value="private"
    >
      <WorldPrivateIcon />
    </KRadio>
  </div>
  <div>Selected: {{ cardRadio || 'none' }}</div>
</div>

```vue
<template>
  <KRadio
    v-model="cardRadio"
    description="Choose this option if you want your APIs to be publicly accessible by anyone on the internet."
    card
    label="Public"
    selected-value="public"
  >
    <WorldIcon />
  </KRadio>
  <KRadio
    v-model="cardRadio"
    description="Choose this option if you want your APIs to only be accessible from within your private network."
    card
    label="Private"
    selected-value="private"
  >
    <WorldPrivateIcon />
  </KRadio>
  <div>Selected: {{ cardRadio || 'none' }}</div>
</template>

<script setup lang="ts">
import { ref } from 'vue'

const cardRadio = ref<string>('')
</script>
```

### labelAttributes

>  KRadio has an instance of KLabel for supporting tooltip text. Use the `labelAttributes` prop to configure the KLabel's [props](/components/label). This example shows using the `labelAttributes` to set up a tooltip. Tooltip content can also be [slotted](#slots).

Author: Done

    </div>
  </template>
</KCard>

```vue
<template>
  <div>Selected: {{ radioValue }}</div>
  <KRadio name="test" :selected-value="true" v-model="radioValue">Boolean</KRadio>
  <KRadio name="test" selected-value="string" v-model="radioValue">String</KRadio>
  <KRadio name="test" :selected-value="objA" v-model="radioValue">Object A</KRadio>
  <KRadio name="test" :selected-value="objB" v-model="radioValue">Object B</KRadio>
</template>

<script setup lang="ts">
import { ref } from 'vue'

const radioValue = ref<boolean | string | object>('string')

const objA = { name: 'a' }
const objB = { name: 'b' }
</script>
```

## Props

### v-model

Use `v-model` to bind the `checked` state of the underlying `<input>` element. The `v-model` binds to the `modelValue` prop of the component and sets the current checked state of the input. You can read more about passing values via `v-model` [here](https://vuejs.org/guide/components/events.html#usage-with-v-model).

### selectedValue

The value of the KRadio option that will be emitted by the `change` and `update:modelValue` events.

### label

Will place label text to the right of the radio. Can also be [slotted](#slots).

> <KRadio v-model="labelPropRadio" label="Label example" :selected-value="true" />

Reviewer: The label and the radio don't appear to be vertically aligned. The label text appears slightly lower than the radio button

![image](https://github.com/Kong/kongponents/assets/2229946/15a55770-2dd9-4e24-b260-21f6b62479a0)

Reviewer: All instances
![image](https://github.com/Kong/kongponents/assets/2229946/18c35cd3-4727-4c3f-881e-c70908760921)

```html
<KRadio v-model="checked" label="Label example" :selected-value="true" />
```

### description

Will place description text under the radio label. Can also be [slotted](#slots).

<KRadio v-model="descriptionPropRadio" label="Label example" description="Some description text" :selected-value="true" />

```html
<KRadio
  v-model="radio"
  label="Label example"
  description="Some description text"
  :selected-value="true"
/>
```

### hasError

Use this prop to apply error styling to the component.

<KRadio v-model="hasErrorPropRadio" label="Input error" has-error description="Some description text" :selected-value="true" />

```html
<KRadio
  v-model="radio"
  label="Input error"
  has-error
  description="Some description text"
  :selected-value="true"
/>
```

### isCard

Set this prop to true to change the appearance of the KRadio component to a card-style design. When `isCard` is set to `true`, the KRadio component will be displayed with a card-like layout, providing a visually distinct and appealing presentation.

<KCard>
  <template #body>

>     <div class="cards-container">

Reviewer: Can we remove the `KCard` wrapper around the radio cards?

# Radio

> KRadio is wrapper for native input type `radio` elements.

Reviewer: ```suggestion
KRadio is a wrapper for native input type `radio` elements.
```

<KCard>

>   <template #title>Selected: {{ radioValue }}</template>

Reviewer: suggestion: can we remove this wrapper KCard? (expect this comment on every docs page; we want to display the components as-is without KCard impacting font sizes, etc.)

### selectedValue

The value of the KRadio option that will be emitted by the `change` and `update:modelValue` events.

### label

Will place label text to the right of the radio. Can also be [slotted](#slots).

<KRadio v-model="labelPropRadio" label="Label example" :selected-value="true" />

```html
<KRadio v-model="checked" label="Label example" :selected-value="true" />
```

### description

Will place description text under the radio label. Can also be [slotted](#slots).

<KRadio v-model="descriptionPropRadio" label="Label example" description="Some description text" :selected-value="true" />

```html
<KRadio
  v-model="radio"
  label="Label example"
  description="Some description text"
  :selected-value="true"
/>
```

### hasError

Use this prop to apply error styling to the component.

<KRadio v-model="hasErrorPropRadio" label="Input error" has-error description="Some description text" :selected-value="true" />

```html
<KRadio
  v-model="radio"
  label="Input error"
  has-error
  description="Some description text"
  :selected-value="true"
/>
```

### isCard

:::danger NOTE

> KRadio card styling is to be revisited in the next `9.0.0-beta.x` release as design is being finalized.

Reviewer: nit: shouldn't this say `alpha`

    <KRadio
      v-model="cardRadio"
      description="Choose this option if you want your APIs to be publicly accessible by anyone on the internet."
      is-card
      label="Public"
      selected-value="public"
    >
      <WorldIcon />
    </KRadio>
    <KRadio
      v-model="cardRadio"
      description="Choose this option if you want your APIs to only be accessible from within your private network."
      is-card
      label="Private"
      selected-value="private"
    >
      <WorldPrivateIcon />
    </KRadio>
  </div>
  <div>Selected: {{ cardRadio }}</div>
</div>

```vue
<template>
  <KRadio
    v-model="cardRadio"
    description="Choose this option if you want your APIs to be publicly accessible by anyone on the internet."
    is-card
    label="Public"
    selected-value="public"
  >
    <WorldIcon />
  </KRadio>
  <KRadio
    v-model="cardRadio"
    description="Choose this option if you want your APIs to only be accessible from within your private network."
    is-card
    label="Private"
    selected-value="private"
  >
    <WorldPrivateIcon />
  </KRadio>
  <div>Selected: {{ cardRadio }}</div>
</template>

<script setup lang="ts">
import { ref } from 'vue'

> const cardRadio = ref<string>('')

Reviewer: nit: Can we display a string of `none` when nothing is selected? (or similar)

</script>
```

### labelAttributes

 KRadio has an instance of KLabel for supporting tooltip text. Use the `labelAttributes` prop to configure the KLabel's [props](/components/label). This example shows using the `labelAttributes` to set up a tooltip. Tooltip content can also be [slotted](#slots).

> <KRadio v-model="labelAttributesPropRadio" label="Tooltips?" :label-attributes="{ info: 'I use the KLabel `help` prop' }" :selected-value="true" />

Reviewer: issue(maybe): if I click on the tooltip icon, it selects the corresponding radio element. Since users may try clicking on the info icon instead of hovering, should we change this behavior to not select the item on click?

## Slots

### default

> Anything passed in to the default slot will replace the `label` prop text.

Reviewer: ```suggestion
Content passed in to the `default` slot will be shown as the label content. The slot content takes precedence over the `label` prop.
```

  Label goes here. The radio is {{ checked ? "selected" : "not selected" }}
</KRadio>
```
:::warning NOTE

> Avoid slotting in block-level elements like `div` into a `default` slot as it will be rendered inside the `label` element. This applies to card-style radio as well.

Reviewer: nit: wording
```suggestion
To preserve a valid HTML structure, avoid slotting in block-level elements such as a `div` into the `default` slot as it will be rendered inside a `label` element. This also applies to card-style radio.
```

Author: Done

:::

:::tip TIP

> When `card` prop is true, the content passed through default slot will render directly above the label. Should you want to customize the layout inside the card you can omit using `label` and `description` props and style content passed through the `default` slot yourself.

Reviewer: ```suggestion
When `card` prop is true, the content passed to the `default` slot will render directly above the label. Should you want to customize the layout inside the card you can omit using the `label` and `description` props and style content passed through the `default` slot yourself.
```

Author: Done

</KRadio>
```

> :::warning NOTE

Reviewer: suggestion: I don't think this needs to be a warning, more just like "info" or "tip"
```suggestion
:::tip NOTE
```

> When `isCard` prop is true, the content passed through default slot will render directly above the label. Should you want to customize the layout inside the card you can omit using `label` and `description` props and style content passed through the `default` slot yourself.

Reviewer: ```suggestion
When `isCard` prop is true, the content passed through default slot will render directly above the label. Should you want to customize the layout inside the card you can omit using `label` and `description` props and style content passed through the `default` slot as desired.
```

:::

### description

> Anything passed in to this slot will replace the `description` prop text.

Reviewer: ```suggestion
Content passed in to the `description` slot will be shown as the description content. The slot content takes precedence over the `description` prop.
```

</KRadio>
```

### tooltip

Provides a slot for tooltip content displayed after the radio label.

<KRadio v-model="tooltipSlotRadio" :selected-value="true">
  My tooltip

>   <template #tooltip>Brings all the <code>devs</code> to the yard</template>

Reviewer: nit: can you change the tooltip text to something diff?

</KRadio>
```

## Events

> KRadio emits two events with same data in payloads.

Reviewer: ```suggestion
KRadio emits two events with same payload.
```

### change

Fired on change, returns the checked status of the radio (`boolean`).

### update:modelValue

> Fired on change, returns the checked status of the radio (`boolean`).

Reviewer: At the top of the docs page in the `selectedValue` docs, we say:

> The value of the KRadio option that will be emitted by the `change` and `update:modelValue` events.

Here in the events section, we don't mention that prop so it's a little confusing as it seems like we're emitting different data. Can we change the wording here?
